ext/mbstring/mbstring.c

```diff
@@ -2277,91 +2278,270 @@ PHP_FUNCTION(mb_strcut)
+	}
+
+	/* The input string is fine; we don't need to append the trim marker */
+	if (from == 0) {
+		GC_ADDREF(input);
+		return input;
+	}
+	return mb_drop_chars(input, enc, from);
+
+	/* The input string is too wide; we need to build a new string which
+	 * includes some portion of the input string, with the trim marker
+	 * concatenated onto it */
+restart_conversion:
+	in = (unsigned char*)ZSTR_VAL(input);
+	in_len = ZSTR_LEN(input);
+	state = 0;
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+
+dont_restart_conversion:
+		if (out_len <= to_skip) {
+			to_skip -= out_len;
+		} else {
+			for (int i = to_skip; i < out_len; i++) {
+				remaining_width -= character_width(wchar_buf[i]);
+				if (remaining_width < 0) {
+					enc->from_wchar(wchar_buf + to_skip, i - to_skip, &buf, true);
+					goto append_trim_marker;
+				}
+			}
+			enc->from_wchar(wchar_buf + to_skip, out_len - to_skip, &buf, !in_len);
+			to_skip = 0;
+		}
+	}
+
+append_trim_marker:
+	if (ZSTR_LEN(marker) > 0) {
+		MB_CONVERT_BUF_ENSURE((&buf), buf.out, buf.limit, ZSTR_LEN(marker));
+		memcpy(buf.out, ZSTR_VAL(marker), ZSTR_LEN(marker));
+		buf.out += ZSTR_LEN(marker);
+	}
+
+	return mb_convert_buf_result(&buf);
+}
+
+/* Trim the string to terminal width; optional, add a 'trim marker' if it was truncated */
 PHP_FUNCTION(mb_strimwidth)
 {
-	char *str, *trimmarker = NULL;
-	zend_string *encoding = NULL;
-	zend_long from, width, swidth = 0;
-	size_t str_len, trimmarker_len;
-	mbfl_string string, result, marker, *ret;
+	zend_string *str, *trimmarker, *encoding = NULL;
+	zend_long from, width;
 
 	ZEND_PARSE_PARAMETERS_START(3, 5)
-		Z_PARAM_STRING(str, str_len)
+		Z_PARAM_STR(str)
 		Z_PARAM_LONG(from)
 		Z_PARAM_LONG(width)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_STRING(trimmarker, trimmarker_len)
+		Z_PARAM_STR(trimmarker)
 		Z_PARAM_STR_OR_NULL(encoding)
 	ZEND_PARSE_PARAMETERS_END();
 
-	string.encoding = marker.encoding = php_mb_get_encoding(encoding, 5);
-	if (!string.encoding) {
+	const mbfl_encoding *enc = php_mb_get_encoding(encoding, 5);
+	if (!enc) {
 		RETURN_THROWS();
 	}
 
-	string.val = (unsigned char *)str;
-	string.len = str_len;
-	marker.val = NULL;
-	marker.len = 0;
-
-	if ((from < 0) || (width < 0)) {
-		swidth = mbfl_strwidth(&string);
-	}
-
-	if (from < 0) {
-		from += swidth;
-	}
-
-	if (from < 0 || (size_t)from > str_len) {
-		zend_argument_value_error(2, "is out of range");
-		RETURN_THROWS();
+	if (from != 0) {
+		size_t str_len = mb_get_strlen(str, enc);
+		if (from < 0) {
+			from += str_len;
+		}
+		if (from < 0 || from > str_len) {
+			zend_argument_value_error(2, "is out of range");
+			RETURN_THROWS();
+		}
 	}
 
 	if (width < 0) {
-		width = swidth + width - from;
-	}
+		width += mb_get_strwidth(str, enc);
 
-	if (width < 0) {
-		zend_argument_value_error(3, "is out of range");
-		RETURN_THROWS();
-	}
+		if (from > 0) {
+			zend_string *trimmed = mb_pick_chars(str, enc, from);
```

Comment:
Allocating a string, just to measure its width and then deallocate it, isn't exactly good for performance, but frankly, using the 'negative width' feature is crazy and probably almost nobody will ever use it.
It's really a misfeature, in my opinion.

Reply:
See docs for `mb_strimwidth`: https://www.php.net/manual/en/function.mb-strimwidth.php

I don't know if we have users who need the negative width feature, but it seems very suspect to me. Anyways, the decision was already made a few years ago so now we need to keep it for BC.

Reply:
Yeah, that [generalization of negative string indexes](https://wiki.php.net/rfc/negative-string-offsets) probably didn't make sense here. It might make sense to deprecate it (probably too late for PHP 8.2, though).

Reply:
@cmb69 That would be *great* if we could deprecate it. It's a very strange feature.

Reply:
OK, will do. Hopefully tomorrow.

Reply:
Edited the RFC which @cmb69 kindly created: https://wiki.php.net/rfc/deprecations_php_8_3

Now I'm wondering if we can also deprecate negative `$from` arguments for `mb_strimwidth` as well.

Reply:
As I understand it, negative `$width`s are expensive to handle, and where not properly implemented, but negative `$start`s shouldn't have these issues, or am I missing something?

Reply:
Negative `$from` was also not properly implemented before this PR. They are also expensive in that they require an extra call to `mb_strlen` which would not be required otherwise.

Reply:
...BUT, I don't have a strong opinion about the idea of deprecating negative `$from`. I was just throwing it out as an idea.